### ISSUE TITLE
GET /realtime/v1/websocket returning 404 while trying to test out examples locally

### ISSUE THREAD

**Issue by bhvngt** [supabase/realtime#263]
# Bug report

## Describe the bug

When I run on docker locally using the [docker-compose](https://github.com/supabase/realtime/blob/master/docker-compose.rls.dev.yml) file provided, I get a stream of following error
```
realtime-realtime-1  | 2022-06-27 17:03:14.003 request_id=FvyI_4WIRrSx3BcAAAHB [info] GET /realtime/v1/websocket
realtime-realtime-1  | 2022-06-27 17:03:14.004 request_id=FvyI_4WIRrSx3BcAAAHB [info] Sent 404 in 275µs
realtime-realtime-1  | 2022-06-27 17:03:18.555 request_id=FvyJAJTVMdBmcLsAAAHh [info] GET /realtime/v1/websocket
realtime-realtime-1  | 2022-06-27 17:03:18.555 request_id=FvyJAJTVMdBmcLsAAAHh [info] Sent 404 in 312µs
```

## To Reproduce

1. git clone https://github.com/supabase/realtime.git
2. cd realtime
3. change `SECURE_CHANNELS: "false"` inside `docker-compose.rls.dev.yml`
4. docker-compose -f docker-compose.rls.dev.yml up --force-recreate --remove-orphans -d
5. cd examples/node-js
6. npm install
7. node src/server.js
```
DatabaseListener connected
SchemaListener connected
TableListener connected
```
8. psql postgresql://postgres:postgres@localhost:5432/postgres
9. INSERT INTO users (name) VALUES ('John');

## Expected behavior

Expecting example code to spit out console logs for inserts

## Screenshots

<img width="1513" alt="image" src="https://user-images.githubusercontent.com/79074469/175998264-81add2c2-7539-404c-989f-1a3d4f255a9c.png">

## System information

- OS: macOS
- Version of supabase-js: using the same version as that of example code.
- Version of Node.js: 18.4

## Additional context

I tried following the fix that was provided in #165. Changing the protocol from `http` to `ws` did not work. `/socket` was already present in the sample code.

I was trying to make this work for my project but it was not able to make any websocket connection. So I thought of making it run locally using the example setup. 

Thanks for this wonderful project. Appreciate your help in fixing this.

**Comment by bhvngt** [comment]
Thanks @w3b6x9 for your quick reply. That did work out fine.

Will the RLS support still exists? As per [this](https://github.com/supabase/realtime#realtime-rls-server-setup), `RLS` is required to enable `RLS` support.

**Comment by w3b6x9** [comment]
@bhvngt well if you disable `SECURE_CHANNELS` then it doesn't require any tokens so RLS won't work. If you want RLS to work you should turn back `SECURE_CHANNELS`.

**Comment by bhvngt** [comment]
ok. So to make it work locally with `RLS`, if I set `SECURE_CHANNELS` to `TRUE` and change URL protocol to `https`, then it does not connect. I am sure I am missing something here. What do I need to do to make it work locally with `RLS` support?

**Comment by w3b6x9** [comment]
@bhvngt ah, gotcha. if you're doing this locally you should continue to use `http` and then pass in `apikey` at a minimum.

```js
import { RealtimeClient } from '@supabase/realtime-js'

var client = new RealtimeClient(process.env.REALTIME_URL, { params: { apikey: 'token123' }})
client.connect()
```

- https://github.com/supabase/realtime-js

where apikey is a generated token from the secret that was defined in `docker-compose.rls.dev.yml`. The token requires a `role`, which is a database role`, to work.

**Comment by bhvngt** [comment]
Cool.. That make sense. 

Does this mean that `REPLICATION_MODE` `RLS` is deprecated and now `STREAM` takes care of both `RLS` and `non-RLS` scenarios? Reason I am highlighting this is because thats what the current documentation suggests.

**Comment by bhvngt** [comment]
@w3b6x9 It does not appear to be working. Here are the steps I followed.

- git pull the latest from the master
- `npm install` to update the dependencies
- changed `REPLICATION_MODE` to `STREAM` in `docker-compose.rls.dev.yml`
- changed `SECURE_CHANNELS` to `false` in `docker-compose.rls.dev.yml`
- generated `JWT` token containing `{"role": "authenticated}` and signed with `JWT_SECRET` `SOMETHING_SUPER_SECRET` present in `docker-compose.rls.dev.yml`
- Changed `src/server.js` with
```
const REALTIME_URL = process.env.REALTIME_URL || 'http://localhost:4000/socket'
const socket = new RealtimeClient(REALTIME_URL, {params: {apikey: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYXV0aGVudGljYXRlZCJ9.jNLynx_4nCpPv1fNkNFOzPQ-Y_vPxJAlDV8sp-590ZM"}})
```
- restarted docker
- ran `ALTER TABLE users ENABLE ROW LEVEL SECURITY` on the database
- `SET ROLE authenticated; SELECT * FROM users` shows zero records to ensure that RLS is working;
- Started node server
- ran `SET ROLE postgres; INSERT INTO users (name) VALUES ('Jean');` on the database
- node server console shows the inserted record.

With RLS enabled for `users` table, I was expecting server to stop showing any inserts. However, it still shows inserts.

Even though I am receiving the realtime feed, I still see, `GET /realtime/v1/websocket` `Sent 404 in 1ms` info logs on the docker realtime logs. Is this expected?

**Comment by w3b6x9** [comment]
> Does this mean that REPLICATION_MODE RLS is deprecated and now STREAM takes care of both RLS and non-RLS scenarios?

@bhvngt it's the opposite. `STREAM` is the initial version of Realtime that doesn't check against RLS policies.

**Comment by w3b6x9** [comment]
@bhvngt try it with `REPLICATION_MODE` set to `RLS` and `SECURE_CHANNELS` set to `true`.

**Comment by bhvngt** [comment]
Thanks a ton @w3b6x9. 

That seems to be working fine. So the missing piece was passing `apikey` from orignal setup, which makes complete sense in the hindsight.

Very much appreciate your help.

As a sidenote, whenever I insert record which is supposed to be filtered due to RLS, I see an error log recorded in the realtime logs. I am assuming that this one is an expected log.

```
realtime-realtime-1  | 2022-06-28 17:51:15.491 [error] %Postgrex.Error{connection_id: 183, message: nil, postgres: %{code: :undefined_function, file: "parse_func.c", hint: "No function matches the given name and argument types. You might need to add explicit type casts.", internal_position: "1", internal_query: "realtime.build_prepared_statement_sql('walrus_rls_stmt', entity_, columns)", line: "629", message: "function realtime.build_prepared_statement_sql(unknown, regclass, realtime.wal_column[]) does not exist", pg_code: "42883", routine: "ParseFuncOrColumn", severity: "ERROR", unknown: "ERROR", where: "PL/pgSQL function realtime.apply_rls(jsonb,integer) line 228 at EXECUTE"}, query: nil}
```

**Comment by w3b6x9** [comment]
@bhvngt

> Thanks a ton @w3b6x9.

You got it!

> whenever I insert record which is supposed to be filtered due to RLS, I see an error log recorded in the realtime logs. I am assuming that this one is an expected log.

That doesn't look right. Looks like it might have missed a migration. Can you try wiping away your docker db and trying again.

**Comment by bhvngt** [comment]
Yep that was the culprit.  Anon volume were still lurking while I restarted my docker-compose run. After removing that the error message disappears. It all look set. 

Thanks once again @w3b6x9 .

**Comment by bhvngt** [comment]
@w3b6x9 Sorry to bother you once again.

I was actually mistaken. RLS with local hosting does not seem to be working. Not sure what I am doing wrong here.

Here are the steps that should reproduce the issue

- Fresh `git pull` of this repo
- Changed  following line inside `examples/node-js/src/server.js` where the `apikey` token is a `jwt` containing `{"role": "authenticated"}` signed with `SOMETHING_SUPER_SECRET`
```
// const socket = new RealtimeClient(REALTIME_URL)
const socket = new RealtimeClient(REALTIME_URL, {params: {apikey: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYXV0aGVudGljYXRlZCJ9.jNLynx_4nCpPv1fNkNFOzPQ-Y_vPxJAlDV8sp-590ZM"}});
```
- Started docker with following command. My guess is that this removes any anon volumes.
` docker-compose -f docker-compose.rls.dev.yml up --force-recreate --remove-orphans -dV -t2`
- Ran `node src/server.js` in another terminal. I get following output.
```
DatabaseListener connected
SchemaListener connected
TableListener connected
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
```
- Opened psql in another terminal and ran following command 
```psql postgresql://postgres:postgres@localhost:5432/postgres```
- Now when I trigger an insert with `INSERT INTO users (name) VALUES ('John');`, `node terminal` correctly logs the change received event.
- However, when I trigger an insert after I enable security with following commands
```
ALTER TABLE users ENABLE ROW LEVEL SECURITY;
INSERT INTO users (name) VALUES ('John');
```
I can see following error reappearing in the `docker realtime` logs
```
realtime-realtime-1  | 2022-07-11 12:00:17.448 [error] %Postgrex.Error{connection_id: 184, message: nil, postgres: %{code: :undefined_function, file: "parse_func.c", hint: "No function matches the given name and argument types. You might need to add explicit type casts.", internal_position: "1", internal_query: "realtime.build_prepared_statement_sql('walrus_rls_stmt', entity_, columns)", line: "629", message: "function realtime.build_prepared_statement_sql(unknown, regclass, realtime.wal_column[]) does not exist", pg_code: "42883", routine: "ParseFuncOrColumn", severity: "ERROR", unknown: "ERROR", where: "PL/pgSQL function realtime.apply_rls(jsonb,integer) line 228 at EXECUTE"}, query: nil}
```

You can find the entire `docker realtime logs` [here](https://gist.github.com/bhvngt/d351c6a19b388e9d365484ca9672c4fa).

In my project, I get similar results.

**Comment by w3b6x9** [comment]
@bhvngt no bother at all, happy to help.

i just had a look at [docker realtime logs](https://gist.github.com/bhvngt/d351c6a19b388e9d365484ca9672c4fa) but didn't see any apikey in the parameters:

```bash
realtime-realtime-1  | 2022-07-11 11:56:11.695 [info] JOINED realtime:public:users in 9ms
realtime-realtime-1  |   Parameters: %{}
```

so something must be wrong with the setup.

which version of `realtime-js` are you running? can you try it with the latest if not already done so.

**Comment by w3b6x9** [comment]
@bhvngt 

```bash
Change received on DatabaseListener {
  columns: [],
  commit_timestamp: null,
  errors: [ 'Error 401: Unauthorized' ],
  record: {},
  schema: 'realtime',
  table: 'subscription',
  type: 'INSERT'
}
```

this is normal b/c you have to explicitly grant select permission for tables outside of the `public` schema.

**Comment by w3b6x9** [comment]
@bhvngt if you haven't already done so can you please `git pull` on this repo's `master` so you have the latest?

**Comment by bhvngt** [comment]
Thanks @w3b6x9 for your help. I pulled the docker image as well as re-cloned this repo. So setup everything from scratch. 

It's still giving the same result. I do notice that the it first connects using the `apikey` and then later while joining `realtime`, `apikey` disappears.
```
2022-07-12 06:05:46.024 [info] CONNECTED TO RealtimeWeb.UserSocket in 215µs
realtime-realtime-1  |   Transport: :websocket
realtime-realtime-1  |   Serializer: Phoenix.Socket.V1.JSONSerializer
realtime-realtime-1  |   Parameters: %{"apikey" => "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYXV0aGVudGljYXRlZCJ9.jNLynx_4nCpPv1fNkNFOzPQ-Y_vPxJAlDV8sp-590ZM", "vsn" => "1.0.0"}
realtime-realtime-1  | 2022-07-12 06:05:46.161 [info] JOINED realtime:* in 133ms
realtime-realtime-1  |   Parameters: %{}
```

Here's the full [server.js](https://gist.github.com/bhvngt/9f71bf9895e3035064420da0b933afdb) that I am running on my setup. Its using the same connection that was used to connect with the realtime server.

**Comment by w3b6x9** [comment]
@bhvngt Ah, I think your JWT needs `exp`. I don't think Realtime will work properly without it.

**Comment by bhvngt** [comment]
@w3b6x9 I have recreated my `jwt` token with following content and secret.

```
{
  "role": "authenticated",
  "exp": 1689144780
}
```
`JWT_SECRET: SOMETHING_SUPER_SECRET`

Then I have updated the token inside `server.js`Token with the generated token `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYXV0aGVudGljYXRlZCIsImV4cCI6MTY4OTE0NDc4MH0.qRMj6BSfgg3a7cz7H57cOnsBJ-mnvl4M-BaW6rNC2W8`

I still get the same result.

While database inserts on tables where `RLS` is `Disabled` works without any issues. It throws error only when `RLS` is `enabled` on the table. 

`realtime` server logs are pretty much same.

I also tried by adding `"iat": 1657608780` into the token. But its still giving the same result.

**Comment by w3b6x9** [comment]
@bhvngt got it!

i've pinged another team member internally regarding:

```bash
realtime-realtime-1  | 2022-07-11 12:00:17.448 [error] %Postgrex.Error{connection_id: 184, message: nil, postgres: %{code: :undefined_function, file: "parse_func.c", hint: "No function matches the given name and argument types. You might need to add explicit type casts.", internal_position: "1", internal_query: "realtime.build_prepared_statement_sql('walrus_rls_stmt', entity_, columns)", line: "629", message: "function realtime.build_prepared_statement_sql(unknown, regclass, realtime.wal_column[]) does not exist", pg_code: "42883", routine: "ParseFuncOrColumn", severity: "ERROR", unknown: "ERROR", where: "PL/pgSQL function realtime.apply_rls(jsonb,integer) line 228 at EXECUTE"}, query: nil}
```

and we'll get back to you shortly on that. sorry about that and thank you for your patience!

**Comment by bhvngt** [comment]
@w3b6x9 Thank you for your wonderful support and building this product.

**Comment by w3b6x9** [comment]
@bhvngt if you want to get unblocked immediately please run everything off of this branch: https://github.com/supabase/realtime/tree/fix/recreate-walrus-function.

**Comment by bhvngt** [comment]
Thanks a ton @w3b6x9. 

After building this locally using `docker-compose build`, now it is working out well. I do see `errors: ['Error 401: Unauthorized']` messages getting logged after couple of seconds. I am assuming that these are expected. 

`realtime server logs` are looking good with zero errors.

Slightly side question - I saw that the realtime publication is enabled on all tables. On a real project, does enabling publication on all tables will have any performance impact assuming that security rules are already in place?

**Comment by w3b6x9** [comment]
> I do see errors: ['Error 401: Unauthorized'] messages getting logged after couple of seconds. I am assuming that these are expected.

@bhvngt yes, these are expected b/c the publication listens to all tables but we only allow clients to listen for changes in `public` schema by default. this opens up the opportunity to listen to tables in private schemas in the future.

> realtime server logs are looking good with zero errors.

@bhvngt sounds great!

> I saw that the realtime publication is enabled on all tables. On a real project, does enabling publication on all tables will have any performance impact assuming that security rules are already in place?

@bhvngt good question. I would say that performance impact is minimal up to a point so I wouldn't worry about that. The bigger worry is security and data egress. All future tables that you add are automatically added to the publication b/c of publication on all tables command so it's easy to forget to turn on RLS. Plus, you'll have data leaving the server that you don't need so you'll only increase your data egress costs.

**Comment by bhvngt** [comment]
> @bhvngt yes, these are expected b/c the publication listens to all tables but we only allow clients to listen for changes in `public` schema by default. this opens up the opportunity to listen to tables in private schemas in the future.

Hmm. In my project I have my data stored in `non-public` schema. Will `realtime` work for tables stored in that schema? If not then is there any workaround possible?

> @bhvngt good question. I would say that performance impact is minimal up to a point so I wouldn't worry about that. The bigger worry is security and data egress. All future tables that you add are automatically added to the publication b/c of publication on all tables command so it's easy to forget to turn on RLS. Plus, you'll have data leaving the server that you don't need so you'll only increase your data egress costs.

Thanks @w3b6x9 Make sense. Will keep this in mind.

**Comment by w3b6x9** [comment]
>  Will realtime work for tables stored in that schema? If not then is there any workaround possible?

@bhvngt so your client will get the change but the actual data is hidden. as superuser you can grant certain roles select access to those tables in the non-public schemas.

don't quote me on this but you can do something like `grant select on all tables in schema [your non-public schema] to authenticated`. then if your clients connect as authenticated role they can get changes from tables in that non-public schema.

**Comment by bhvngt** [comment]
Thank you so much @w3b6x9. This was very helpful indeed. I checked and it works like you said. Will not quote you though :)

I will continue to use the build from [fix/recreate-walrus-function](https://github.com/supabase/realtime/tree/fix/recreate-walrus-function). I am assuming that sometime in the future this will get merged to master docker build `(supabase/realtime)` hosted on `dockerhub`.

**Comment by w3b6x9** [comment]
@bhvngt haha if it's working then please quote me 😉 

i think it's been deployed already. try https://hub.docker.com/layers/realtime/supabase/realtime/v0.22.7/images/sha256-a96ddfa6b7e95334b22c1bab04df7d22a5bcd98c1ac0ab265616765019df9fc7.

**Comment by bhvngt** [comment]
> i think it's been deployed already. try https://hub.docker.com/layers/realtime/supabase/realtime/v0.22.7/images/sha256-a96ddfa6b7e95334b22c1bab04df7d22a5bcd98c1ac0ab265616765019df9fc7.

Super 🚀. Love you guys for your help and support.
